launcher/minecraft/ComponentUpdateTask.cpp

```diff
@@ -512,24 +531,198 @@ void ComponentUpdateTask::resolveDependencies(bool checkOnly)
     if (recursionNeeded) {
         loadComponents();
     } else {
+        finalizeComponents();
         emitSucceeded();
     }
 }
 
+// Variant visitation via lambda
+template <class... Ts>
+struct overload : Ts... {
+    using Ts::operator()...;
+};
+template <class... Ts>
+overload(Ts...) -> overload<Ts...>;
+
+void ComponentUpdateTask::performUpdateActions()
+{
+    auto& instance = d->m_profile->d->m_instance;
+    bool addedActions;
+    QStringList toRemove;
+    do {
+        addedActions = false;
+        toRemove.clear();
+        auto& components = d->m_profile->d->components;
+        auto& componentIndex = d->m_profile->d->componentIndex;
+        for (auto component : components) {
+            if (!component) {
+                continue;
+            }
+            auto action = component->getUpdateAction();
+            auto visitor = overload{
+                [](const UpdateActionNone&) {
+                    // noop
+                },
+                [&component, &instance](const UpdateActionChangeVersion& cv) {
+                    qCDebug(instanceProfileResolveC) << instance->name() << "|" << "UpdateActionChangeVersion" << component->getID() << ":"
+                                                     << component->getVersion() << "change to" << cv.targetVersion;
+                    component->setVersion(cv.targetVersion);
+                    component->waitLoadMeta();
+                },
+                [&component, &instance](const UpdateActionLatestRecommendedCompatable lrc) {
+                    qCDebug(instanceProfileResolveC)
+                        << instance->name() << "|" << "UpdateActionLatestRecommendedCompatable" << component->getID() << ":"
+                        << component->getVersion() << "updating to latest recommend or compatible with" << lrc.parentUid << lrc.version;
+                    auto versionList = APPLICATION->metadataIndex()->get(component->getID());
+                    versionList->waitToLoad();
+                    if (versionList) {
```

Comment:
You already access the pointer before verifying it: this may cause segmentation faults.
Either switch the lines or disregard the if as the get should always return a non-nullptr versionList.